mri/reconstructors/base.py

```diff
@@ -0,0 +1,146 @@
+        elif fourier_type == 'cartesian':
+            self.fourier_op = FFT(
+                samples=kspace_loc,
+                shape=uniform_data_shape,
+                n_coils=n_coils)
+        elif fourier_type == 'stack':
+            self.fourier_op = Stacked3DNFFT(kspace_loc=kspace_loc,
+                                            shape=uniform_data_shape,
+                                            implementation=nfft_implementation,
+                                            n_coils=n_coils)
+        else:
+            raise ValueError('The value of fourier_type must be "cartesian" | '
+                             '"non-cartesian" | "stack"')
+        if verbose >= 5:
+            print("Initialized fourier operator : " + str(self.fourier_op))
+
+    def reconstruct(self, x_init=None):
+        """ This method calculates operator transform.
+        Parameters
+        ----------
+        x_init: np.ndarray (optional, default None)
+            input initial guess image for reconstruction
+        """
+        raise NotImplementedError("'reconstruct' is an abstract method.")
+
+
+class ReconstructorWaveletBase(ReconstructorBase):
+    """ This is the base reconstructor class for reconstruction.
+    This class holds some common parameters that is common for all
+    MR Image reconstructors
+    Parameters
+    ----------
+    kspace_loc: np.ndarray
+        the mask samples in the Fourier domain.
+    uniform_data_shape: tuplet
+        the shape of the matrix containing the uniform data.
+    n_coils: int, default 1
+        Number of coils used to acquire the signal in case of multiarray
+        receiver coils acquisition. If n_coils > 1, data shape must be
+        [n_coils, *data_shape]
+    fourier_type: str (optional, default 'cartesian')
+        type of fourier operator : 'cartesian' | 'non-cartesian' | 'stack'
+    nfft_implementation: str, default 'cpu'
+        way to implement NFFT : 'cpu' | 'cuda' | 'opencl'
+    wavelet_name: str | int
+        if implementation is with waveletN the wavelet name to be used during
+        the decomposition, else implementation with waveletUD2 where the
+        wavelet name is wavelet_id Refer to help of mr_transform under option
+        '-t' to choose the right wavelet_id.
+    padding_mode: str (optional, default zero)
+        The padding mode used in the Wavelet transform,
+        'zero' | 'periodization'
+    nb_scale: int (optional default is 4)
+        the number of scale in the used by the multi-scale wavelet
+        decomposition
+    wavelet_op_per_channel: bool
+        whether wavelet transform should be applied on every channel. This is
+        True only for Calibrationless reconstruction
+    n_jobs: int, default 1
+        Number of parallel jobs for linear operation
+    verbose: int
+        verbosity level for debug, please check dervied class for details
+        on verbosity levels
+    """
+    def __init__(self, kspace_loc, uniform_data_shape, n_coils,
```

Comment:
let me see if we are on same page:
You are asking me to do as follows:

```python
def __init__(self, fourier_type, **kwargs):
        # Define the linear/fourier operators
        if fourier_type == 'non-cartesian':
            self.fourier_op = NonCartesianFFT(**kwargs)
```

??

While I agree that this is surely much easier to document and maintain, doesnt this actually make the code less readable?

Reply:
Why do you think it makes the code less readable? Maybe you mean the docs?

If you look at libraries like for example `matplotlib`, they have to do this all the time because their objects are so deeply derived or reused, they can't afford to explicitly expose all the arguments to every function.
Look at for example the docs to the `plot` function: https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.plot.html
You can see that they have a huge amount of `kwargs` (here the docs are super nice because they also  list the `kwargs` directly and automatially but generally it's fine to just point to the corresponding function or class). This eases the maintaining of the docs and code because when you change a function somewhere you don't need to change the docs and code everywhere.